### ISSUE TITLE
Pulseaudio playback doesn't work under low PULSE_LATENCY_MSEC due to buffer underflow hang

### ISSUE THREAD

**Issue by qsniyg** [libsdl-org/SDL#4387]
~~I apologize for the vague title, I'm not actually even sure what is causing this on my end. Any SDL application (including [this test app](https://gist.github.com/armornick/3447121)) fails to play audio, and this has been the case for the past couple of years (not even sure when this started). However, every other application I've tried works without issue (browsers, vlc, mpv, wine, openal-based applications, etc.). In fact, since I rarely use SDL applications, I disregarded this as being unimportant. I've tried disabling and enabling various soundcards, but this doesn't appear to cause any difference.~~

Edit: I have now discovered that the root cause is due to `PULSE_LATENCY_MSEC` being set to a low value (30) under my configuration, in order to reduce lag. While disabling it (or setting it to a higher value, such as 1000) does fix the issue, it would be nice if SDL were to support lower latencies too :)

This issue is specific to pulseaudio. When pulseaudio is not running, SDL applications play audio without issue. Note that all other applications that work appear to be using pulseaudio directly as well without issue. Strangely, setting `SDL_AUDIODRIVER=alsa` when running any SDL applications also fails if pulseaudio is running. (Edit: alsa works when latency is set to a higher value)

Adding some traces into the source code, it manages to run `PULSEAUDIO_PlayDevice` once, then it runs  `PULSEAUDIO_WaitDevice` a few (4-8) times (without running `PULSEAUDIO_PlayDevice`) and hangs during `PULSEAUDIO_pa_mainloop_iterate` (specifically `pa_mainloop_poll`) the last time. I'm curious if this is related to https://github.com/libsdl-org/SDL/issues/3996 in some way.

`strace` shows similar results:

```
[pid 2] ppoll([{fd=8, events=POLLIN}, {fd=11, events=POLLIN}, {fd=12, events=POLLIN}], 3, NULL, NULL, 8 <unfinished ...>
[pid 1] <... clock_nanosleep resumed>0x7ffe399cfd20) = 0
[pid 1] clock_nanosleep(CLOCK_REALTIME, 0, {tv_sec=0, tv_nsec=100000000}, 0x7ffe399cfd20) = 0
[pid 1] clock_nanosleep(CLOCK_REALTIME, 0, {tv_sec=0, tv_nsec=100000000}, 0x7ffe399cfd20) = 0
...
```

A hacky workaround is to reduce the minimum writable size, for example:

```
        if (PULSEAUDIO_pa_stream_writable_size(h->stream) >= h->mixlen/8) {
            return;
        }
```

This appears to work fine and doesn't crash anything (no audible crackles either). However this is obviously not a good solution. Looking into pacat's source code (simplest official example app I could find), it looks like they handle writing to the stream using callbacks: https://github.com/pulseaudio/pulseaudio/blob/1a73fb9eb9851d00e2907a3dbd50d8dfb5e173b2/src/utils/pacat.c#L458

`pa_stream_set_write_callback(stream, stream_write_callback, NULL);`

This of course doesn't match SDL's internal architecture, but perhaps this could be implemented in WaitDevice by waiting on a semaphore? It appears to be the "official" way to do it.

Wine on the other hand works by dynamically sleeping before writing to the stream: https://github.com/wine-mirror/wine/blob/ebd344f2922f4044117904e024a0a87576a3eff1/dlls/winepulse.drv/mmdevdrv.c#L527

Both work fine under my configuration.

---

To give a bit more information on mixlen, the chosen SDL_AudioSpec for the test app is:

```
format: AUDIO_S16LSB
size: 8192
channels: 2
samples: 2048 (originally 4096, but /2'd by PULSEAUDIO_OpenDevice)
freq: 44100
```

When writable_size > 0, it always finishes with a writable_size of 3968 before hanging (with ~1770 and ~3700 before, the exact values change from run to run).

**Comment by kcgen** [comment]
Reproduced a similar hang-on-audio-device close due to waiting for a thread. 

Start by removing/moving the ALSA conf files:

```shell
sudo mv /usr/share/alsa/alsa.conf /usr/share/alsa/alsa.conf-bak
sudo mv /usr/share/alsa/alsa.conf.d /usr/share/alsa/alsa.conf.d-bak
sudo mv /etc/alsa/conf.d /etc/alsa/conf.d-bak
```

```c++
#include <stdio.h>
#include <stdlib.h>

#include <SDL2/SDL.h>

int main(int argc, char** argv) {

    SDL_Log("Initializing SDL");
    if (SDL_Init(SDL_INIT_EVERYTHING) != 0) {
        fprintf(stderr, "Failed to initialize SDL:  %s\n", SDL_GetError());
        return 1;
    }

    SDL_Log("Registering SDL_Quit() to run at-exit");
    atexit(SDL_Quit);

    SDL_AudioSpec want;
    SDL_zero(want);

    want.freq = 48000;
    want.format = AUDIO_S16;
    want.channels = 2;
    want.samples = 1024;
    want.callback = NULL;

    SDL_AudioSpec have;

    SDL_Log("Opening the audio device");
    SDL_AudioDeviceID dev = SDL_OpenAudioDevice(
        NULL, 0, &want, &have, SDL_AUDIO_ALLOW_FORMAT_CHANGE);

    // Failure check
    if (dev == 0) {
        SDL_Log("Failed to open the audio device: %s", SDL_GetError());
        return 1;
    }

    // Report format changes
    if (have.format != want.format) {
            SDL_Log("Received a different audio format (OK!)");
    }

    SDL_Log("Unpausing the audio device");
    SDL_PauseAudioDevice(dev, 0);

    SDL_Log("Playing 3 seconds of silence");
    SDL_Delay(3000);

    SDL_Log("Pausing the audio device");
    SDL_PauseAudioDevice(dev, 1);

    SDL_Log("Closing the audio device");
    SDL_CloseAudioDevice(dev);

    SDL_Log("Exiting");
    return 0;
}
```

Save that to `hangtest.c`, then: `g++ -g test_sdl.c -lSDL2 -o hangtest`

![2021-09-11_17-49](https://user-images.githubusercontent.com/1557255/132967106-f5c6f06c-46b8-4c17-8528-7476973826a4.png)
`<hang!>`

Thanks to @pellcorp for the report and @arrowgent for assisting.  Reported here: https://github.com/dosbox-staging/dosbox-staging/issues/1252

**Comment by icculus** [comment]
> A hacky workaround is to reduce the minimum writable size, for example:

This is probably what we should do for 2.0.18 to mitigate the issue, and then look at reworking this to use callbacks in 2.0.20.

**Comment by icculus** [comment]
Okay, the "wait for a small portion of the mixbuffer to be available" workaround is in there now for 2.0.18, moving this issue to 2.0.20 for more heavy lifting, like moving to Pulse's callback API.
